File: 1335-maximum-candies-allocated-to-k-children/maximum-candies-allocated-to-k-children.py
class Solution:
    def maximumCandies(self, candies: List[int], k: int) -> int:

        def good(val):
            if val == 0:
                return True
            tot = 0
            for candy in candies:
                tot += candy // val
            # print(tot, val)
            return tot >= k

        # print(good(5))

        right = sum(candies) // k
        left = 0
        if right == 0:
            return 0

        while left < right:
            mid = (left + right) // 2
            if good(mid):
                left = mid + 1
            else:
                right = mid
        # print(mid)
        if good(mid+1): return mid+1
        if good(mid): return mid
        if good(mid-1): return mid-1
        return mid

File: 1335-maximum-candies-allocated-to-k-children/maximum-candies-allocated-to-k-children.cpp
class Solution {
public:
    int maximumCandies(vector<int>& candies, long long k) {

        auto good = [&] (int val) {
            long long count = 0;
            if (val == 0) return true;
            for (int i = 0; i < candies.size(); ++i){
                count += (candies[i] / val);
                if (count >= k) return true;
            }
            return false;
        };

        int l = 1;
        int r = accumulate(candies.begin(), candies.end(), 0LL) / k;
        int res = 0;

        while (l <= r) {
            int mid = (l+r) / 2;
            if (good(mid)) {
                res = mid;
                l = mid+1;
            }
            else {
                r = mid-1;
            }
        }

        return res;
        
    }
};